### DISCUSSION TITLE
provide example for babel-standalone for data-plugins

### DISCUSSION THREAD

**Discussion by ghost** [babel/babel#15464] [Q&A]
Hello,

I'm trying to have a  build system free project, and babel standalone works flawless, I'm looking for documentation for using plugins through the data-plugin.

https://github.com/babel/babel/discussions/12059 I looked up an existing answer here but it says to use a build system.

**Comment by JLHwung** [comment]
We do have an example of custom plugins passed by `data-plugin`: https://babel.dev/docs/babel-standalone#custom-plugins

**Reply by ghost** [reply]
Registering a plug-in from javascript is must?

**Reply by JLHwung** [reply]
Yes. `babel-standalone` must be aware of such plugin.

**Comment by ghost** [comment]
For some consumers who may not know much about javascript find difficult or in my case, I was hoping adding just the markup with data attributes be enough. 


Do you see its a good idea to move js implementation out from consumers and let babel automatically pull details out from data-plugins values?

**Reply by ghost** [reply]
Example - automatic registration of plug-in by detecting the data-plugin attribute

**Reply by JLHwung** [reply]
If you are using plugins included in `babel-standalone`, you can specify that in `data-plugins`. e.g.

```html
<script type="text/babel" data-plugins="proposal-optional-chaining">
```

> babel automatically pull details out from data-plugins values?

If you are using custom plugins, the `Babel.registerPlugin` API is exactly for this purpose. Babel has to keep a map from data-plugin values to the actual plugin definition otherwise it can't automatically know which plugin you are using.
